### PR TITLE
MEJO-18 Adjust LIDAR interface to provide meaningful data

### PR DESCRIPTION
**Pull request by plainte**

### CODE REVIEW

File: Nucleo_F746ZG_app/Src/rplidar_driver_interface.c [hunk 1]
```diff
@@ -54,3 +54,27 @@ void rplidar_reset()
...
+  int tmp_size = (angle_to - angle_from) + 1;
+  int temp[tmp_size];
+  if (tmp_size > 1)
+  {
+    for (int i = 0; i < tmp_size; i++)
+    {
+      temp[i] = rplidar_distance_data[i + angle_from];
+    }
+    bubble_sort(temp, tmp_size);
+    int median = get_median(temp, tmp_size);
```
Comment (pVer297) [inline Nucleo_F746ZG_app/Src/rplidar_driver_interface.c]: You should implement the bubble sort inside the get_median function. Without it the get_median is not doing what the name suggests

File: Nucleo_F746ZG_app/.cproject [hunk 2]
```diff
@@ -229,6 +242,19 @@
...
+						<tool customBuildStep="true" id="org.eclipse.cdt.managedbuilder.ui.rcbs.982092014" name="Resource Custom Build Step">
+							<inputType id="org.eclipse.cdt.managedbuilder.ui.rcbs.inputtype.1928344095" name="Resource Custom Build Step Input Type">
+								<additionalInput kind="additionalinputdependency" paths="" />
+							</inputType>
+							<outputType id="org.eclipse.cdt.managedbuilder.ui.rcbs.outputtype.526376854" name="Resource Custom Build Step Output Type" />
+						</tool>
+					</fileInfo>
 					<sourceEntries>
 						<entry flags="VALUE_WORKSPACE_PATH|RESOLVED" kind="sourcePath" name="startup" />
 						<entry flags="VALUE_WORKSPACE_PATH|RESOLVED" kind="sourcePath" name="Middlewares" />
```
[the remark this answers is not in the document]
Comment (plainte) [reply Nucleo_F746ZG_app/.cproject]: -u_print_float changed it probably

File: Nucleo_F746ZG_app/.cproject [hunk 3]
```diff
@@ -53,7 +53,7 @@
 									<listOptionValue builtIn="false" value="STM32F746xx" />
```
Comment (johnyka) [inline Nucleo_F746ZG_app/.cproject]: I don't understand most of these changes. What did you change in the project file?

File: Nucleo_F746ZG_app/Src/test.c [hunk 4]
```diff
@@ -37,25 +31,40 @@ void TestTask(void const * argument)
...
     BSP_LED_On(LED_RED);
     BSP_LED_On(LED_BLUE);
-    osDelay(100);
-    BSP_LED_Off(LED_GREEN);
-    BSP_LED_Off(LED_RED);
-    BSP_LED_Off(LED_BLUE);
-    osDelay(100);
-    osThreadTerminate(testHandle);
+    osDelay(500);
+    servo_turn_continous(LEFT, 100, 5);    //left 100% turn very slow
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/test.c]: Why did you put servo tests here? And why did you not put lidar tests?

Comment (plainte) [reply Nucleo_F746ZG_app/Src/test.c]: it wasn't me i just pulled the master

File: Nucleo_F746ZG_app/.cproject (hunk 3 above)
Comment (plainte) [reply Nucleo_F746ZG_app/.cproject]: -u_print_float

File: Nucleo_F746ZG_app/Src/rplidar_interrupt_interface.c [hunk 5]
```diff
@@ -9,8 +9,10 @@ void rplidar_receive_device_health_data()
 void rplidar_parse_device_health_data()
 {
   device_health_data.status = get_health_data_received[0];
-  device_health_data.error_code = ((uint16_t) get_health_data_received[1] << 8)
-      | get_health_data_received[2];
+  device_health_data.error_code = ((uint16_t) get_health_data_received[1] << 8) | get_health_data_received[2];
+  char check[10];
+  sprintf(check, "%d", device_health_data.status);
+  HAL_UART_Transmit(&huart3, (uint8_t*)check, strlen(check), 0xFFFF);
```
Comment (pVer297) [inline Nucleo_F746ZG_app/Src/rplidar_interrupt_interface.c]: async_log

File: Nucleo_F746ZG_app/Src/test.c [hunk 6]
```diff
@@ -79,7 +78,37 @@ void TestTask(void const * argument)
       sprintf(rgb_number, "R%d G%d B%d C%d \r\n", rgb_colors(RED),rgb_colors(GREEN), rgb_colors(BLUE), rgb_colors(CLEAR));
       HAL_UART_Transmit(&huart3, (uint8_t*) rgb_number, strlen(rgb_number),0xFFFF);
       osDelay(500);
+    }*/
```
Comment (pVer297) [inline Nucleo_F746ZG_app/Src/test.c]: you should remove the /**/ from the pull request so the master has a working semi-auto test

File: Nucleo_F746ZG_app/Src/utility.c [hunk 7]
```diff
@@ -2,7 +2,7 @@
 
 void utility_init()
 {
-  osMailQDef(mail_box, 16, messages_t);
+  osMailQDef(mail_box, 50, messages_t);
```
Comment (pVer297) [inline Nucleo_F746ZG_app/Src/utility.c]: why have you changed this?
it was approved? (this is 2.5 kbyte of memory for async_log)

File: Nucleo_F746ZG_app/.cproject (hunk 2 above)
Comment (pVer297) [inline Nucleo_F746ZG_app/.cproject]: I still think we should not change this. *Waiting for tech lead review*

File: Nucleo_F746ZG_app/Src/test.c (hunk 6 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/test.c]: its working

File: Nucleo_F746ZG_app/Src/utility.c (hunk 7 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/utility.c]: coz i need it

File: Nucleo_F746ZG_app/Src/rplidar_driver_interface.c [hunk 8]
```diff
@@ -54,3 +57,51 @@ void rplidar_reset()
...
+{
+  int tmp_size = 0;
+  if (angle_from < angle_to) {
+    tmp_size = (angle_to - angle_from) + 1;
+  } else if (angle_from == angle_to) {
+    tmp_size = 1;
+  } else if (angle_from > angle_to) {
+    tmp_size = angle_to + 360 - angle_from + 1;
+  }
+  int *temp = (int*) pvPortMalloc(tmp_size * sizeof(int));
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rplidar_driver_interface.c]: if this function can be called from UART then it's not going to work. Can you solve the uart command without calling this function?

File: Nucleo_F746ZG_app/test/test_rplidar_calc.c [hunk 9]
```diff
@@ -0,0 +1,44 @@
+#include "unity.h"
```
Comment (johnyka) [inline Nucleo_F746ZG_app/test/test_rplidar_calc.c]: I twould be nice to test the functionality of get_range_distance. The problem is that pvPortMalloc won't work here. I could suggest to use a 360 array on the stack and only use what's necessary. It's a little bit wasteful I know, but it solves the problem for both of my comments.

Comment (plainte) [reply Nucleo_F746ZG_app/test/test_rplidar_calc.c]: done

File: Nucleo_F746ZG_app/Src/rplidar_driver_interface.c (hunk 8 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/rplidar_driver_interface.c]: done

File: Nucleo_F746ZG_app/Src/utility.c (hunk 7 above)
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/utility.c]: I just suggested it for testing purposes, I don't think we need that much.

File: Nucleo_F746ZG_app/Src/rplidar_utility.c [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+    }
+    median = get_median(temp, tmp_size);
+  } else {
+    if (angle_from == 360) {
+      median = rplidar_distance_data[0];
+    } else {
+      median = rplidar_distance_data[angle_from];
+    }
+  }
+  //async_log("distance %d - %d: %d\r\n", angle_from, angle_to, median);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rplidar_utility.c]: pls remove this line instead of commenting

File: Nucleo_F746ZG_app/Src/interface_execute.c [hunk 11]
```diff
@@ -66,16 +66,20 @@ void execute_lidar(command_t* command)
...
   case LIDAR_START:
     rplidar_start_motor();
+    rplidar_start_scan();
     break;
   case LIDAR_STOP:
     rplidar_stop_motor();
     break;
   case LIDAR_READ_CONTINOUS:
-    rplidar_start_scan();
+    //task to be implemented
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/interface_execute.c]: actually that's part of this task :)

File: Nucleo_F746ZG_app/Src/interface_execute.c [hunk 12]
```diff
@@ -66,16 +66,20 @@ void execute_lidar(command_t* command)
...
     break;
   case LIDAR_READ_STOP:
     rplidar_stop_scan();
     break;
+  case LIDAR_READ_SINGLE:
+    rplidar_get_distance(command->params.lidar.value, command->params.lidar.value2);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/interface_execute.c]: The returned value should be printed out

File: Nucleo_F746ZG_app/Src/utility.c (hunk 7 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/utility.c]: done

File: Nucleo_F746ZG_app/Src/rplidar_utility.c (hunk 10 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/rplidar_utility.c]: done

File: Nucleo_F746ZG_app/Src/interface_execute.c (hunk 12 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/interface_execute.c]: done

File: Nucleo_F746ZG_app/Src/interface_execute.c (hunk 11 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/interface_execute.c]: done master

File: Nucleo_F746ZG_app/Src/rplidar_driver_interface.c [hunk 13]
```diff
@@ -3,6 +3,9 @@
 void rplidar_start_motor()
 {
   HAL_GPIO_WritePin(GPIOB, LIDAR_MOTCTRL_Pin, GPIO_PIN_SET);
+  for (int i = 0; i < 360; i++) {
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rplidar_driver_interface.c]: a memset would be a little bit better.

File: Nucleo_F746ZG_app/Src/usart.c [hunk 14]
```diff
@@ -159,13 +159,13 @@ void HAL_UART_MspInit(UART_HandleTypeDef* uartHandle)
...
     GPIO_InitStruct.Mode = GPIO_MODE_AF_PP;
-    GPIO_InitStruct.Pull = GPIO_NOPULL;
+    GPIO_InitStruct.Pull = GPIO_PULLUP;
     GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_VERY_HIGH;
     GPIO_InitStruct.Alternate = GPIO_AF7_USART2;
     HAL_GPIO_Init(GPIOD, &GPIO_InitStruct);
 
     /* USART2 interrupt Init */
-    HAL_NVIC_SetPriority(USART2_IRQn, 5, 0);
+    HAL_NVIC_SetPriority(USART2_IRQn, 0, 0);
```
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/usart.c]: this should not have 0 priority. There are much more important interrupts than this one

File: Nucleo_F746ZG_app/Src/rplidar_driver_interface.c (hunk 13 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/rplidar_driver_interface.c]: done

File: Nucleo_F746ZG_app/Src/usart.c (hunk 14 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/usart.c]: done

File: Nucleo_F746ZG_app/Src/rplidar_utility.c (hunk 10 above)
Comment (johnyka) [inline Nucleo_F746ZG_app/Src/rplidar_utility.c]: it's not done :)

**Comment by johnyka** [comment]
Pls remove that comment that you forgot to and I can approve

File: Nucleo_F746ZG_app/Src/rplidar_utility.c (hunk 10 above)
Comment (plainte) [reply Nucleo_F746ZG_app/Src/rplidar_utility.c]: doone

**Pull request merged**